client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java

```diff
@@ -163,6 +239,90 @@ public void unregisterShuffle(int shuffleId) {
     shuffleStatus.remove(shuffleId);
   }
 
+  private static class ShuffleServerFailureRecord {
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final ReentrantReadWriteLock.ReadLock readLock = lock.readLock();
+    private final ReentrantReadWriteLock.WriteLock writeLock = lock.writeLock();
+    private final Map<String, AtomicInteger> shuffleServerFailureRecordCount;
+    private int stageAttemptNumber;
+
+    private ShuffleServerFailureRecord(
+        Map<String, AtomicInteger> shuffleServerFailureRecordCount, int stageAttemptNumber) {
+      this.shuffleServerFailureRecordCount = shuffleServerFailureRecordCount;
+      this.stageAttemptNumber = stageAttemptNumber;
+    }
+
+    private <T> T withReadLock(Supplier<T> fn) {
```

Comment:
These methods seems useful. Could we extract some methods?

Reply:
> These methods seems useful. Could we extract some methods for other classes to use?

Later PR will use this method.